libyul/backends/evm/EVMCodeTransform.cpp

```diff
@@ -635,6 +680,45 @@ void CodeTransform::visitStatements(vector<Statement> const& _statements)
 	for (auto const& statement: _statements)
 	{
 		freeUnusedVariables();
+
+		if (!m_delayedReturnVariables.empty() && (
+			holds_alternative<VariableDeclaration>(statement) ||
+			holds_alternative<Leave>(statement) ||
+			holds_alternative<ForLoop>(statement) ||
+			holds_alternative<Block>(statement) ||
+			holds_alternative<Switch>(statement) ||
+			holds_alternative<If>(statement) ||
+			[&](){
+				ReferencesCounter referencesCounter{ReferencesCounter::CountWhat::OnlyVariables};
+				referencesCounter.visit(statement);
+				return ranges::any_of(
+					m_delayedReturnVariables,
+					[&](YulString _returnVariableName) { return referencesCounter.references().count(_returnVariableName); }
+				);
+			}()
+		))
+		{
+			for (YulString _returnVariableName: m_delayedReturnVariables)
+			{
+				auto* var = std::get_if<Scope::Variable>(m_scope->lookup(_returnVariableName));
+				yulAssert(var, "Return variable not in scope.");
+				m_context->variableStackHeights[var] = static_cast<size_t>(m_assembly.stackHeight());
+				// Preset stack slots for return variables to zero.
+				m_assembly.appendConstant(u256(0));
+				if (!m_unusedStackSlots.empty())
```

Comment:
Is there a test for this case?

Reply:
It's gone with the next commit anyways. This what was triggered the comment about return variables not being contiguous, i.e. they idea here was to interleave them with arguments into the unused argument slots... but that didn't really work and in the end I just plainly push all of them contiguously after the last argument again (only potentially popping arguments at the top), which keeps things simpler.